File: backend/app/schemas/knowledge.py
from datetime import datetime
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel


class CollectionOut(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    color: str
    documentCount: int
    createdAt: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm(cls, col) -> "CollectionOut":
        return cls(
            id=col.id,
            name=col.name,
            description=col.description,
            color=col.color,
            documentCount=len(col.documents) if col.documents else 0,
            createdAt=col.created_at,
        )


class CreateCollectionRequest(BaseModel):
    name: str
    description: Optional[str] = None
    color: Optional[str] = "#6366f1"


class DocumentOut(BaseModel):
    id: UUID
    name: str
    type: str
    size: int
    status: str
    collectionId: Optional[UUID] = None
    tags: List[str] = []
    pageCount: Optional[int] = None
    wordCount: Optional[int] = None
    createdAt: datetime
    updatedAt: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm(cls, doc) -> "DocumentOut":
        return cls(
            id=doc.id,
            name=doc.name,
            type=doc.file_type,
            size=doc.file_size,
            status=doc.status.value if hasattr(doc.status, "value") else doc.status,
            collectionId=doc.collection_id,
            tags=doc.tags or [],
            pageCount=doc.page_count,
            wordCount=doc.word_count,
            createdAt=doc.created_at,
            updatedAt=doc.updated_at,
        )


class ChunkOut(BaseModel):
    id: UUID
    documentId: UUID
    content: str
    chunkIndex: int
    tokenCount: Optional[int] = None
    createdAt: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm(cls, chunk) -> "ChunkOut":
        return cls(
            id=chunk.id,
            documentId=chunk.document_id,
            content=chunk.content,
            chunkIndex=chunk.chunk_index,
            tokenCount=chunk.token_count,
            createdAt=chunk.created_at,
        )


class ConnectorOut(BaseModel):
    id: UUID
    type: str
    name: str
    status: str
    lastSyncAt: Optional[datetime] = None
    documentCount: int
    config: Any

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm(cls, connector) -> "ConnectorOut":
        return cls(
            id=connector.id,
            type=connector.type,
            name=connector.name,
            status=connector.status.value if hasattr(connector.status, "value") else connector.status,
            lastSyncAt=connector.last_sync_at,
            documentCount=connector.document_count,
            config=connector.config or {},
        )


class CreateConnectorRequest(BaseModel):
    type: str
    name: str
    config: Any = {}


class KnowledgeStats(BaseModel):
    totalDocuments: int
    totalChunks: int
    storageUsedBytes: int
    totalCollections: int
    totalConnectors: int
    lastIndexedAt: Optional[datetime] = None

File: frontend/src/types/knowledge.ts
export type ConnectorType =
  | 'google_drive'
  | 'sharepoint'
  | 'onedrive'
  | 'confluence'
  | 'notion'
  | 'jira'
  | 'github'
  | 'gitlab'
  | 'slack'
  | 'teams'
  | 'salesforce'
  | 'hubspot'
  | 'zendesk'
  | 'intercom'
  | 'dropbox'
  | 'box'
  | 'figma'
  | 'linear'
  | 'asana'
  | 'trello'
  | 'airtable'
  | 'web_scraper'
  | 'rss_feed'
  | 's3';

export type SyncStatus = 'idle' | 'syncing' | 'success' | 'error' | 'paused';

export type DocumentType =
  | 'pdf'
  | 'docx'
  | 'doc'
  | 'pptx'
  | 'xlsx'
  | 'txt'
  | 'md'
  | 'html'
  | 'csv'
  | 'json'
  | 'image'
  | 'video'
  | 'audio';

export type ProcessingStatus = 'pending' | 'processing' | 'indexed' | 'failed';

export interface Document {
  id: string;
  name: string;
  type: DocumentType;
  size: number;
  mimeType: string;
  url?: string;
  thumbnailUrl?: string;
  collectionId?: string;
  collectionName?: string;
  connectorId?: string;
  connectorType?: ConnectorType;
  processingStatus: ProcessingStatus;
  chunkCount: number;
  tokenCount: number;
  language?: string;
  tags: string[];
  metadata: Record<string, unknown>;
  uploadedBy: string;
  createdAt: string;
  updatedAt: string;
  lastSyncedAt?: string;
}

export interface Chunk {
  id: string;
  documentId: string;
  content: string;
  embedding?: number[];
  pageNumber?: number;
  position: number;
  tokenCount: number;
  metadata: Record<string, unknown>;
}

export interface Collection {
  id: string;
  name: string;
  description?: string;
  color: string;
  icon?: string;
  documentCount: number;
  totalSize: number;
  isPublic: boolean;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface Connector {
  id: string;
  type: ConnectorType;
  name: string;
  description: string;
  logoUrl: string;
  status: 'connected' | 'disconnected' | 'error' | 'pending';
  syncStatus: SyncStatus;
  lastSyncedAt?: string;
  nextSyncAt?: string;
  documentCount: number;
  errorMessage?: string;
  config: Record<string, unknown>;
  createdAt: string;
}

export interface KnowledgeStats {
  totalDocuments: number;
  totalChunks: number;
  totalTokens: number;
  storageUsedBytes: number;
  storageQuotaBytes: number;
  connectorCount: number;
  activeConnectors: number;
  lastUpdatedAt: string;
}

export interface UploadProgress {
  fileId: string;
  fileName: string;
  progress: number;
  status: 'queued' | 'uploading' | 'processing' | 'done' | 'error';
  errorMessage?: string;
}
